lib/gollum-lib/filter/yaml.rb

```diff
@@ -0,0 +1,26 @@
+require 'yaml'
+# Extract YAML frontmatter from data and build metadata table. 
+
+class Gollum::Filter::YAML < Gollum::Filter
+  
+  # Regexp thanks to jekyll
+  YAML_FRONT_MATTER_REGEXP = %r!\A(---\s*\n.*?\n?)^((---|\.\.\.)\s*$\n?)!m
+
+  def extract(data)
+    data.gsub!(YAML_FRONT_MATTER_REGEXP) do
+      @markup.metadata ||= {}
+      begin
+        frontmatter = ::YAML.load(Regexp.last_match[1])
+        frontmatter = frontmatter.respond_to?(:to_h) ? frontmatter.to_h : Hash.new(frontmatter)
```

Comment:
I think what threw me was `Hash.new(frontmatter)` as opposed to just have it be `Hash.new`. The Array edge case is interesting, since `to_h` is defined on each array, but will break on a normal array of type `['a', 'b']` with a `TypeError`. So we would need to rescue those if we want to continue using `to_h`. Alternatively, we could check if the YAML loaded object responds to `keys` and `values` methods?